src/adaptors.rs

```diff
@@ -972,3 +974,55 @@ impl<I> Iterator for Combinations<I> where I: Iterator + Clone, I::Item: Clone{
         size_hint::add((lo / 2, hi.map(|hi| hi / 2)), extra)
     }
 }
+
+/// An iterator adapter to filter non-unique elements.
+///
+/// See [*.unique()*](trait.Itertools.html#method.unique) for more information.
+#[derive(Clone)]
+pub struct UniqueBy<I: Iterator, V, F> {
+    iter: I,
+    used: HashSet<V>,
+    f: F,
+}
+
+impl<I: Iterator, V, F> UniqueBy<I, V, F> where V: Clone + Eq + Hash, F: FnMut(&I::Item) -> V {
+    /// Create a new **UniqueBy** iterator.
+    pub fn new(iter: I, f: F) -> UniqueBy<I, V, F> {
+        UniqueBy {
+            iter: iter,
+            used: HashSet::new(),
+            f: f,
+        }
+    }
+}
+
+impl<I, V, F> Iterator for UniqueBy<I, V, F> where
+    I: Iterator,
+    V: Clone + Eq + Hash,
+    F: FnMut(&I::Item) -> V
+{
+    type Item = I::Item;
+
+    fn next(&mut self) -> Option<I::Item> {
+        loop {
+            match self.iter.next() {
+                None => return None,
+                Some(v) => {
+                    let key = (self.f)(&v);
+                    if self.used.insert(key) {
+                        return Some(v);
+                    }
+                }
+            }
+        }
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.iter.size_hint()
```

Comment:
This should use the upper bound of the adapted iterator, and use a minimum of 0. Unless the hashset is empty, then use either 0 or 1 as lower bound depending on the adapted itertator.

Simply put -- our full information about how many elements remain, keeping in mind that they may all be duplicates or all unique.